### ISSUE TITLE
Editor jumps around when images rendered in markdown editor

### ISSUE THREAD

**Issue by shbach** [laurent22/joplin#5918]
## Environment

Joplin version: 2.6.10
Platform: Linux/Windows
OS specifics: Tried on both Solus Linux and Windows 10

## Steps to reproduce

1. Install rich markdown plugin and enable "Render images below their markdown source" option.
2. Open a long note with images (can use [example long.jex](https://discourse.joplinapp.org/t/plugin-rich-markdown/15053/281))
3. Scroll around the note and see the view jump.

![450944439a67a34808be80f59d61e76820c015f3](https://user-images.githubusercontent.com/5298745/147418703-a04c40b1-2233-4cbe-a81f-79b725f2f5e3.gif)

## Describe what you expected to happen

Originally reported [here](https://discourse.joplinapp.org/t/plugin-rich-markdown/15053/283). The editor shouldn't jump around like this and is quite disruptive to Joplin usage.

**Comment by laurent22** [comment]
Is that something that should be fixed on the app or on the plugin?

**Comment by CalebJohn** [comment]
@laurent22 I'm think it's the app because there is no scroll handling in the plugin as of v0.8.1. Unfortunately, I can't reproduce this behaviour, so I can't be certain.

**Comment by ken1kob** [comment]
I've almost figured out the cause. Fixing the app is needed.

**Comment by ken1kob** [comment]
I inspected the problem.

Details of the problem:
- When the 3000msec interval timer of Rich Markdown plugin is triggered, an old scroll position is sometimes restored.
- The problem happens only if image reandering is enabled, because the timer is used only in the case.

The problem is caused by a complicated interplay among Joplin's sync scroll, Rich Markdown plugin, and CodeMirror. 

Causes:
- Joplin's sync scroll restores an old scroll position when resized.
- The stored scroll position is updated by scroll events except ones raised by resizing, since scroll events raised by resizing are ignored.
- CodeMirror has no separate resize event. It has `refresh` event raised by resizing and refreshing. The cause of the event cannot be distinguished.
- Rich Markdown's interval timer calls CodeMirror.refresh() several times for each interval. CodeMirror.refresh() raises `refresh` event.
- If the timer is triggered when scrolling, scroll events at the time are considered "after resizing", and the stored scroll position is not updated.
- When the next timer is triggered, it is considered as resized, and the non-updated old stored scroll position is restored. 

Planned Treatments:
1. [Joplin] Updates a stored scroll position for each scroll event regardless of whether raised by resizing or not.
2. [Joplin] Distinguishes resizing `refresh` events as possible.
3. (Optional) [Rich Markdown] Reduces CodeMirror.refresh() calls.

**Comment by ken1kob** [comment]
@CalebJohn, I think a modification of Rich Markdown is maybe optional.
But, there are some unclear points for me.

In Rich Markdown's refreshAllWidgets():
```js
    wid.node.onload = function() {
        let im = this as HTMLElement;
        if (im.clientWidth != width || im.clientHeight != height) {
            cm.refresh();
        }
    };
```

- Since clientHeight/clientWidth property of image elements are usually zero, cm.refresh() is almost always called. Is this as you intended?
- cm.refresh() is called the same number of times as the number of images in a note. Can you reduce these calls?

**Comment by CalebJohn** [comment]
Thank you for digging into this @ken1kob 

> - Since clientHeight/clientWidth property of image elements are usually zero, cm.refresh() is almost always called. Is this as you intended?

That is not intended! It's a mistake of me not testing this feature on a robust test note. I never tested on long notes where some of the images would be scrolled out of view, which results in them having a client[Width/Height] of zero. 
I've just put out an update now (https://github.com/CalebJohn/joplin-rich-markdown/commit/b97f0ad9f1ed0176c97e4b5871424d944ec55ead) that checks for the condition where the image isn't visible and skips this check. The result is that `cm.refresh` is only called when an image was changed on disk resulting in a new image size (which was the original intention with this code).
Thanks for finding that issue @ken1kob, I appreciate your help on this.

> - cm.refresh() is called the same number of times as the number of images in a note. Can you reduce these calls?

For now the above change should be sufficiently reduce these calls. With v0.8.2 this will only be called once for each image that changes size on disk. Ideally it would only be called once even if all visible images are changes, but I don't think that's worth the trade-off in complexity.

---

After my changes, this should be functionally fixed for the users (refresh won't be called at all for most users). I think it's still worth fixing this on the Joplin end if it can be done in a way that doesn't add too much complexity. Otherwise, I think we should document it somewhere that calling refresh can confuse the scrolling. Plugins really shouldn't be calling refresh that often, and doing so in rich-markdown plugin was actually a pretty bad bug.

**Comment by ken1kob** [comment]
@CalebJohn , I agree. Now, I'm trying to fix the problem in the app side too.

BTW, I tried Rich Markdown 0.8.2. It still periodically calls CodeMirror.refresh() with less frequency.
Is it OK to refer clientWidth/Height? How about width/height or naturalWidth/Height?

**Comment by CalebJohn** [comment]
@ken1kob Thanks for letting me know about that. I learned today how many different size measurements there are on the DOM, I've fixed the plugin now by using just width/height instead of clientWidth/Height. I believe this will fix all remaining refresh calls. I spent some time reviewing how I handle those resize events because I don't like manually checking every N seconds. If you have any ideas/tips I'd like to hear them!

**Comment by ken1kob** [comment]
PR #5929 will fix the problem from the app side. It mitigates the symptom happened with Rich Markdown 0.8.1. With RM 0.8.3, the symptom is not observed during my tests.

One minor limitation still remains.
CodeMirror.refresh() redoes the layout of a note, and the result and its height slightly differs depending on a scroll position. So, when CodeMirror.refresh() is called during scrolling, a small scroll jump (1 to several pixels) sometimes happens.
This is due to CodeMirror, and I have not found a workaround yet. Fortunately, its impact is small, and it will rarely happen in RM 0.8.3+.

**Comment by ken1kob** [comment]
> I spent some time reviewing how I handle those resize events because I don't like manually checking every N seconds. If you have any ideas/tips I'd like to hear them!

I think the current one is not bad. I don't know an efficient way to watch changes of local and remote resources specified by URL without an external push notification service.

**Comment by laurent22** [comment]
> I don't know an efficient way to watch changes of local and remote resources specified by URL without an external push notification service.

Would it help if we add a new plugin API like `onResourceChange(resourceId)`?

**Comment by CalebJohn** [comment]
> Would it help if we add a new plugin API like `onResourceChange(resourceId)`?

Yes, in that case I'd be able to hook in to the event and only refresh the images that have changed.

**Comment by laurent22** [comment]
`onResouceChange` is now available from this commit: https://github.com/laurent22/joplin/commit/2660ff3af6232588de9092e3f8fdca68a06e8f5e It takes one object that has an `id` property (the resource ID).

**Comment by CalebJohn** [comment]
> `onResouceChange` is now available from this commit: [2660ff3](https://github.com/laurent22/joplin/commit/2660ff3af6232588de9092e3f8fdca68a06e8f5e) It takes one object that has an `id` property (the resource ID).

Thanks @laurent22 that implementation is simpler than I had expected! I'll get this hooked up for the next release!
